kitsune/middleware.py

```diff
@@ -0,0 +1,199 @@
+# -*- coding: utf-8 -*-
+#
+# This tries to fix bug 1357563 in a best effort way. It may find a proper page
+# to redirect users to but if it fails it will raise a 404.
+#
+# Flow:
+#
+# 1. Check if the response is 404 and it's from a Lithium URL (/t5/*)
+# 2. Checks cache, redirects if successful
+# 3. Checks Lithium localized product_pages from LithiumRedirectionMiddleware.product_pages, redirects if successful
+# 4. Fetches the page from Lithium and tries to extract title of the page.
+#    Use og:title instead of HTML Title because Lithium strips the later down to 50 chars or something.
+#    If it fails it returns 404
+# 5a. Try to find a Document with the same title, redirect if successful
+# 5b. Try to find a Topic with the same title under a product. Product is also extracted by title. For non-English pages we need to do a reverse translation lookup from TOPICS
+# 5c. Try to find a Product. For non-English pages we need to do a reverse translation lookup from PRODUCTS
+# 5d. If everything fails redirect to the search page with title as search string.
```

Comment:
Good catch, we talked with ben and agreed to remove this to track the failures. updating the docs.